In [60]:
import pandas as pd
import matplotlib.pyplot as plt
import joblib



In [61]:
# import scikit learn
from sklearn.model_selection import train_test_split
import seaborn as sns

In [62]:
import sklearn
print(sklearn.__version__)

1.5.1


# `2_display setting`

In [63]:
sklearn.set_config(transform_output="pandas")

## `3_Getting a Data`

In [64]:
flights = pd.read_csv('./data/flight_price.csv')
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [65]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


### Primilary Analysis

### check data types

In [66]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

### checking for duplicates

In [67]:
flights.duplicated().sum()

220

In [68]:
(
    flights
    .loc[df.duplicated(keep=False)]
    .sort_values(['Airline','Date_of_Journey','Source','Destination'])
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### changing the type of date time column

In [69]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [70]:
flights.sample()

Airline Date_of_Journey   Source Destination      Route Dep_Time  \
184  IndiGo      12/05/2019  Chennai     Kolkata  MAA → CCU    14:45   

    Arrival_Time Duration Total_Stops Additional_Info  Price  
184        17:05   2h 20m    non-stop         No info   3858

## `Observation`


-  The date of journey should be converted to datetime format
-  The type `Duration `is mixed . It should be numeric type.
-  Total stop should also be numeric type.
-  There are 220 duplicates needs to drop .


## `3_Detailed Analysis`

#### AirLine 

In [71]:
flights['Airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [72]:
flights['Airline'].value_counts()

Airline
Jet Airways                          3849
IndiGo                               2053
Air India                            1752
Multiple carriers                    1196
SpiceJet                              818
Vistara                               479
Air Asia                              319
GoAir                                 194
Multiple carriers Premium economy      13
Jet Airways Business                    6
Vistara Premium economy                 3
Trujet                                  1
Name: count, dtype: int64

- some of enteries have inconsistent values

In [73]:
(
    flights
    .Airline
    .str.replace('Premium economy' ,'')
    .str.replace('Business' ,'')
    .str.title() # to make every word capital after space
    .unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Vistara ',
       'Jet Airways ', 'Multiple Carriers ', 'Trujet'], dtype=object)

## Dep_Time

In [74]:
# to get the values other than numbers

(
    flights
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

In [75]:
df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [76]:
pd.to_datetime(flights.Dep_Time).dt.time

C:\Users\user\AppData\Local\Temp\ipykernel_7036\1056054517.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flights.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

#### Arrival Time

In [77]:
(
    flights
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    .str.split(" ",n=1)
    .str.get(1)
    .unique()
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

#### Total_Stops

In [78]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [79]:
(
    flights
	.Total_Stops
	.replace("non-stop", "0")
	.str.replace(" stops?", "", regex=True) # " stop" " stops"
	.pipe(lambda ser: pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

### function of converting into numbers

In [82]:
def convert_to_minutes(ser):
	return (
		ser
		.str.split(" ", expand=True)
		.set_axis(["hour", "minute"], axis=1)
		.assign(
			hour=lambda df_: (
				df_
				.hour
				.str.replace("h", "")
				.astype(int)
				.mul(60)
			),
			minute=lambda df_: (
				df_
				.minute
				.str.replace("m", "")
				.fillna("0")
				.astype(int)
			)
		)
		.sum(axis=1)
	)

## `4_Data Cleaning`

In [83]:
def clean_data(df):
	return (
		df
		.drop(index=[6474])
		.drop_duplicates()
		.assign(**{
			col: df[col].str.strip()
			for col in df.select_dtypes(include="O").columns
		})
		.rename(columns=str.lower)
		.assign(
			airline=lambda df_: (
				df_
				.airline
				.str.replace(" Premium economy", "")
				.str.replace(" Business", "")
				.str.title()
			),
			date_of_journey=lambda df_: pd.to_datetime(df_.date_of_journey, dayfirst=True),
			dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
			arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,
			duration=lambda df_: df_.duration.pipe(convert_to_minutes),
			total_stops=lambda df_: (
				df_
				.total_stops
				.replace("non-stop", "0")
				.str.replace(" stops?", "", regex=True)
				.pipe(lambda ser: pd.to_numeric(ser))
			),
			additional_info=lambda df_: df_.additional_info.replace("No info", "No Info")
		)
		.drop(columns="route")
	)

In [87]:
flights_cleaned = clean_data(flights)
flights_cleaned

C:\Users\user\AppData\Local\Temp\ipykernel_7036\3708629844.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time=lambda df_: pd.to_datetime(df_.dep_time).dt.time,
C:\Users\user\AppData\Local\Temp\ipykernel_7036\3708629844.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time=lambda df_: pd.to_datetime(df_.arrival_time).dt.time,


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## `5_Split the Data`

In [88]:
flights_final = flights_cleaned.sample(1000)

In [89]:
X = flights_final.drop(columns="price")
y = flights_final.price.copy()

In [90]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(640, 9) (640,)
(160, 9) (160,)
(200, 9) (200,)


## `6_Export the subset`

In [97]:
import pandas as pd
import os

def export_data(X, y, name):
    # Create the filename
    file_name = f"{name}.csv"
    
    # Use the current working directory
    file_path = os.path.join(os.getcwd(), file_name)

    # Combine X and y, and export to CSV
    combined_data = X.join(y)
    combined_data.to_csv(file_path, index=False)

    # Return the first few rows of the exported data
    return pd.read_csv(file_path).head()

In [100]:
export_data(X_train, y_train, "train")

airline date_of_journey   source destination  dep_time  \
0             Indigo      2019-03-27   Mumbai   Hyderabad  19:05:00   
1        Jet Airways      2019-05-06  Kolkata    Banglore  14:05:00   
2             Indigo      2019-03-21  Kolkata    Banglore  21:25:00   
3  Multiple Carriers      2019-03-27    Delhi      Cochin  09:45:00   
4           Air Asia      2019-06-09  Kolkata    Banglore  15:10:00   

  arrival_time  duration  total_stops additional_info  price  
0     20:30:00        85          0.0         No Info   2754  
1     19:50:00       345          1.0         No Info  14781  
2     00:05:00       160          0.0         No Info   4148  
3     22:20:00       755          1.0         No Info   6601  
4     23:30:00       500          2.0         No Info   7870

In [98]:
export_data(X_val, y_val, "val")

airline date_of_journey    source destination  dep_time  \
0            Vistara      2019-03-24   Kolkata    Banglore  20:20:00   
1          Air India      2019-06-24     Delhi      Cochin  07:05:00   
2  Multiple Carriers      2019-06-15     Delhi      Cochin  08:30:00   
3           Spicejet      2019-05-15  Banglore       Delhi  21:10:00   
4          Air India      2019-03-27     Delhi      Cochin  09:45:00   

  arrival_time  duration  total_stops additional_info  price  
0     18:45:00      1345          1.0         No Info  12102  
1     09:25:00      1580          2.0         No Info  11185  
2     19:15:00       645          1.0         No Info   8937  
3     00:05:00       175          0.0         No Info   3625  
4     09:25:00      1420          2.0         No Info   7090

In [99]:
export_data(X_test, y_test, "test")

airline date_of_journey    source destination  dep_time  \
0  Multiple Carriers      2019-06-01     Delhi      Cochin  09:45:00   
1             Indigo      2019-05-12  Banglore       Delhi  16:55:00   
2            Vistara      2019-03-01  Banglore   New Delhi  19:40:00   
3        Jet Airways      2019-03-01  Banglore   New Delhi  16:55:00   
4  Multiple Carriers      2019-06-27     Delhi      Cochin  07:00:00   

  arrival_time  duration  total_stops additional_info  price  
0     16:10:00       385          1.0         No Info   9646  
1     19:55:00       180          0.0         No Info   4823  
2     22:20:00       160          0.0         No Info  21730  
3     10:25:00      1050          1.0  1 Long layover  25735  
4     19:00:00       720          1.0         No Info  12192